File: Examples/ControlPointDemo_ObjC/ControlPointDemo/Player.m
//
//  Player.m
//
//  Copyright (c) 2015 David Robles
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

#import "Player.h"
@import UPnAtom;

typedef NS_ENUM(NSInteger, PlayerState) {
    PlayerStateUnknown = 0,
    PlayerStateStopped,
    PlayerStatePlaying,
    PlayerStatePaused
};

@interface Player ()
@property (nonatomic, readwrite) NSArray *playlist;
@property (nonatomic, readwrite) UIBarButtonItem *playPauseButton;
@property (nonatomic, readwrite) UIBarButtonItem *stopButton;
@property (nonatomic) PlayerState playerState;
@end

@implementation Player {
    NSInteger _position;
    id _avTransportEventObserver;
    NSString *_avTransportInstanceID;
}

+ (Player *)sharedInstance {
    static Player *instance = nil;
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
        instance = [Player new];
        
        instance.playPauseButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"play_button"] style:UIBarButtonItemStylePlain target:instance action:@selector(playPauseButtonTapped:)];
        instance.stopButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"stop_button"] style:UIBarButtonItemStylePlain target:instance action:@selector(stopButtonTapped:)];
        instance->_avTransportInstanceID = @"0";
    });
    
    return instance;
}

- (void)setMediaRenderer:(MediaRenderer1Device *)mediaRenderer {
    MediaRenderer1Device *oldRenderer = _mediaRenderer;
    _mediaRenderer = mediaRenderer;
    
    if (_avTransportEventObserver != nil) {
        [oldRenderer.avTransportService removeEventObserver:_avTransportEventObserver];
    }
    
    _avTransportEventObserver = [mediaRenderer.avTransportService addEventObserver:[NSOperationQueue currentQueue] callBackBlock:^(UPnPEvent *event) {
        if ([event.service isAVTransport1Service] && [event isAVTransport1Event]) {
            AVTransport1Event *avTransportEvent = (AVTransport1Event *)event;
            NSLog(@"%@ Event: %@", event.service.className, avTransportEvent.instanceState);
            NSString *transportState = [avTransportEvent.instanceState[@"TransportState"] lowercaseString];
            if (transportState.length) {
                if ([transportState rangeOfString:@"playing"].location != NSNotFound) {
                    self.playerState = PlayerStatePlaying;
                }
                else if ([transportState rangeOfString:@"paused"].location != NSNotFound) {
                    self.playerState = PlayerStatePaused;
                }
                else if ([transportState rangeOfString:@"stopped"].location != NSNotFound) {
                    self.playerState = PlayerStateStopped;
                }
                else {
                    self.playerState = PlayerStateUnknown;
                }
            }
        }
    }];
}

- (void)startPlayback:(NSArray *)playList position:(NSInteger)position{
    [self setPlaylist:playList];
    
    [self startPlayback:position];
}

- (void)startPlayback:(NSInteger)position{
    //Do we have a Renderer & a playlist ?
    if(_mediaRenderer == nil || _playlist == nil){
        return;
    }
    
    if(position >= [_playlist count]){
        position = 0; //Loop
    }
    
    _position = position;

    ContentDirectory1Object *item = _playlist[position];
    //Is it a Media1ServerItem ?
    if([item isContentDirectory1VideoItem]){
        ContentDirectory1VideoItem *item = _playlist[position];
        
        NSString *uri = [item resourceURL].absoluteString;
        if (uri.length) {
            __weak typeof(self) weakSelf = self;
            [self.mediaRenderer.avTransportService setAVTransportURIWithInstanceID:_avTransportInstanceID currentURI:uri currentURIMetadata:@"" success:^{
                NSLog(@"URI Set succeeded!");
                
                [weakSelf playWithSuccess:^{
                    NSLog(@"Play command succeeded!");
                } failure:^(NSError *error) {
                    NSLog(@"Play command failed: %@", error.localizedDescription);
                }];
            } failure:^(NSError *error) {
                NSLog(@"URI Set failed: %@", error.localizedDescription);
            }];
        }
    }
}

#pragma mark - Internal lib

- (void)playPauseButtonTapped:(id)sender {
    switch (self.playerState) {
        case PlayerStatePlaying:
            [self pauseWithSuccess:^{
                NSLog(@"Pause command succeeded!");
            } failure:^(NSError *error) {
                NSLog(@"Pause command failed: %@", error.localizedDescription);
            }];
            break;
            
        case PlayerStatePaused:
            [self playWithSuccess:^{
                NSLog(@"Play command succeeded!");
            } failure:^(NSError *error) {
                NSLog(@"Play command failed: %@", error.localizedDescription);
            }];
            break;
            
        case PlayerStateStopped:
            [self playWithSuccess:^{
                NSLog(@"Play command succeeded!");
            } failure:^(NSError *error) {
                NSLog(@"Play command failed: %@", error.localizedDescription);
            }];
            break;
            
        default:
            NSLog(@"Play/Pause button cannot be used in this state.");
            break;
    }
}

- (void)stopButtonTapped:(id)sender {
    switch (self.playerState) {
        case PlayerStatePlaying:
            [self stopWithSuccess:^{
                NSLog(@"Stop command succeeded!");
            } failure:^(NSError *error) {
                NSLog(@"Stop command failed: %@", error.localizedDescription);
            }];
            break;
            
        case PlayerStatePaused:
            [self stopWithSuccess:^{
                NSLog(@"Stop command succeeded!");
            } failure:^(NSError *error) {
                NSLog(@"Stop command failed: %@", error.localizedDescription);
            }];
            break;
            
        case PlayerStateStopped:
            NSLog(@"Stop button cannot be used in this state.");
            break;
            
        default:
            NSLog(@"Stop button cannot be used in this state.");
            break;
    }
}

- (void)setPlayerState:(PlayerState)playerState {
    _playerState = playerState;
    
    switch (playerState) {
        case PlayerStateStopped:
            self.playPauseButton.image = [UIImage imageNamed:@"play_button"];
            break;
            
        case PlayerStatePlaying:
            self.playPauseButton.image = [UIImage imageNamed:@"pause_button"];
            break;
            
        case PlayerStatePaused:
            self.playPauseButton.image = [UIImage imageNamed:@"play_button"];
            break;
            
        default:
            self.playPauseButton.image = [UIImage imageNamed:@"play_button"];
            break;
    }
}

- (void)playWithSuccess:(void (^)(void))success failure:(void (^)(NSError *))failure {
    [self.mediaRenderer.avTransportService playWithInstanceID:_avTransportInstanceID speed:@"1" success:success failure:failure];
}

- (void)pauseWithSuccess:(void (^)(void))success failure:(void (^)(NSError *))failure {
    [self.mediaRenderer.avTransportService pauseWithInstanceID:_avTransportInstanceID success:success failure:failure];
}

- (void)stopWithSuccess:(void (^)(void))success failure:(void (^)(NSError *))failure {
    [self.mediaRenderer.avTransportService stopWithInstanceID:_avTransportInstanceID success:success failure:failure];
}

@end

File: Examples/ControlPointDemo_Swift/ControlPointDemo/Player.swift
//
//  Player.swift
//
//  Copyright (c) 2015 David Robles
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

import UIKit
import UPnAtom

private let _PlayerSharedInstance = Player()

class Player {
    class var sharedInstance: Player {
        return _PlayerSharedInstance
    }
    var mediaServer: MediaServer1Device?
    var mediaRenderer: MediaRenderer1Device? {
        didSet {
            didSetRenderer(oldRenderer: oldValue, newRenderer: mediaRenderer)
        }
    }
    fileprivate(set) var playPauseButton: UIBarButtonItem! // TODO: Should ideally be a constant, see Github issue #10
    fileprivate(set) var stopButton: UIBarButtonItem! // TODO: Should ideally be a constant, see Github issue #10
    
    fileprivate var _position: Int = 0
    fileprivate var _playlist: [ContentDirectory1Object]?
    fileprivate var _avTransportEventObserver: AnyObject?
    fileprivate var _playerState: PlayerState = PlayerState.stopped {
        didSet {
            playerStateDidChange()
        }
    }
    fileprivate var _avTransportInstanceID = "0"
    
    enum PlayerState {
        case unknown
        case stopped
        case playing
        case paused
    }
    
    init() {
        playPauseButton = UIBarButtonItem(image: UIImage(named: "play_button"), style: .plain, target: self, action: #selector(Player.playPauseButtonTapped(_:)))
        stopButton = UIBarButtonItem(image: UIImage(named: "stop_button"), style: .plain, target: self, action: #selector(Player.stopButtonTapped(_:)))
    }
    
    func startPlayback(_ playlist: [ContentDirectory1Object], position: Int) {
        _playlist = playlist
        
        startPlayback(position: position)
    }
    
    func startPlayback(position: Int) {
        _position = position
        
        if let item = _playlist?[position] as? ContentDirectory1VideoItem {
            let uri = item.resourceURL.absoluteString
            let instanceID = _avTransportInstanceID
            mediaRenderer?.avTransportService?.setAVTransportURI(instanceID: instanceID, currentURI: uri, currentURIMetadata: "", success: { () -> Void in
                print("URI set succeeded!")
                self.play({ () -> Void in
                    print("Play command succeeded!")
                    }, failure: { (error) -> Void in
                        print("Play command failed: \(error)")
                })
                
                }, failure: { (error) -> Void in
                    print("URI set failed: \(error)")
            })
        }
    }
    
    @objc fileprivate func playPauseButtonTapped(_ sender: AnyObject) {
        print("play/pause button tapped")
        
        switch _playerState {
        case .playing:
            pause({ () -> Void in
                print("Pause command succeeded!")
            }, failure: { (error) -> Void in
                print("Pause command failed: \(error)")
            })
        case .paused, .stopped:
            play({ () -> Void in
                print("Play command succeeded!")
                }, failure: { (error) -> Void in
                    print("Play command failed: \(error)")
            })
        default:
            print("Play/pause button cannot be used in this state.")
        }
    }
    
    @objc fileprivate func stopButtonTapped(_ sender: AnyObject) {
        print("stop button tapped")
        
        switch _playerState {
        case .playing, .paused:
            stop({ () -> Void in
                print("Stop command succeeded!")
                }, failure: { (error) -> Void in
                    print("Stop command failed: \(error)")
            })
        case .stopped:
            print("Stop button cannot be used in this state.")
        default:
            print("Stop button cannot be used in this state.")
        }
    }
    
    fileprivate func didSetRenderer(oldRenderer: MediaRenderer1Device?, newRenderer: MediaRenderer1Device?) {
        if let avTransportEventObserver: AnyObject = _avTransportEventObserver {
            oldRenderer?.avTransportService?.removeEventObserver(avTransportEventObserver)
        }
        
        _avTransportEventObserver = newRenderer?.avTransportService?.addEventObserver(OperationQueue.current, callBackBlock: { (event: UPnPEvent) -> Void in
            if let avTransportEvent = event as? AVTransport1Event,
                let transportState = (avTransportEvent.instanceState["TransportState"] as? String)?.lowercased() {
                    print("\(String(describing: event.service?.className)) Event: \(avTransportEvent.instanceState)")
                    print("transport state: \(transportState)")
                    if transportState.range(of: "playing") != nil {
                        self._playerState = .playing
                    }
                    else if transportState.range(of:"paused") != nil {
                        self._playerState = .paused
                    }
                    else if transportState.range(of:"stopped") != nil {
                        self._playerState = .stopped
                    }
                    else {
                        self._playerState = .unknown
                    }
            }
        })
    }
    
    fileprivate func playerStateDidChange() {
        switch _playerState {
        case .stopped, .paused, .unknown:
            playPauseButton.image = UIImage(named: "play_button")
        case .playing:
            playPauseButton.image = UIImage(named: "pause_button")
        }
    }
    
    fileprivate func play(_ success: @escaping () -> Void, failure:@escaping (_ error: NSError) -> Void) {
        self.mediaRenderer?.avTransportService?.play(instanceID: _avTransportInstanceID, speed: "1", success: success, failure: failure)
    }
    
    fileprivate func pause(_ success: @escaping () -> Void, failure:@escaping (_ error: NSError) -> Void) {
        self.mediaRenderer?.avTransportService?.pause(instanceID: _avTransportInstanceID, success: success, failure: failure)
    }
    
    fileprivate func stop(_ success: @escaping () -> Void, failure:@escaping (_ error: NSError) -> Void) {
        self.mediaRenderer?.avTransportService?.stop(instanceID: _avTransportInstanceID, success: success, failure: failure)
    }
}
